ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNode.java

```diff
@@ -69,16 +65,11 @@ public KsqlStructuredDataOutputNode(
     this.keyField = Objects.requireNonNull(keyField, "keyField")
         .validateKeyExistsIn(schema);
     this.ksqlTopic = Objects.requireNonNull(ksqlTopic, "ksqlTopic");
-    this.outputProperties = Objects.requireNonNull(outputProperties, "outputProperties");
+    this.partitionBy = partitionBy;
```

Comment:
I think the current code is harder to understand than what was there before. Now I need to know that the keyField that is being passed in is _after_ resolving the partition by whereas beforehand I didn't. I think we can remedy this issue by changing `partitionBy` to `needsRepartition` and adding a comment explaining that if `needsRepartition` is set, then the source key field is different from the output key field.

Another option is just performing the check for needsRepartition here instead of in logical planner.

Reply:
I can rename to `needsRepartition`.  

As for it being harder to understand... well, the only code required both a keyField and some properties. The only thing it used the properties for is to get the `PARTITION BY` field from.  If present, the constructor checked that the `PARTITION BY` field matches the `keyField` and throws if it doesn't. Later, when building the query it again gets the `PARTITION BY` field from the properties, (the one we know matches the keyField), and uses it in `selectKey`.

The new code just takes a boolean to indicate the repartition is needed. 

Personally, I see this as a _lot_ simpler.

> Now I need to know that the keyField that is being passed in is after resolving the partition by whereas beforehand I didn't.

The keyField hasn't changed. The keyField has always been the keyField of the resulting stream i.e. the key field of what is returned by `buildStream`, i.e. _after_ the partition by is applied.

Reply:
Have another look and let me know if there is anything else to do to make it easier to understand.

However, I *think* your view that its harder to understand may have come about due to you thinking the value passed to keyField has changed.  The logic remains the same, I've just removed the duplicate data and the checks that were in place to ensure the data was aligned.

Reply:
Why are you up at 2AM!? But anyway, I see your point. I agree now that this is simpler... every time I look at this class I get confused (I remember reviewing your previous PR that fixed some of the partition by stuff and being quite confused).

Reply:
Yeah, at some point I would like to find the time to split or refactor this class.  Each of these `node` classes are a mix of the logical query model and builders of the physical.